src/network/connection.cpp

```diff
@@ -0,0 +1,589 @@
+		int packetSize = m_UDP.parsePacket();
+		if (!packetSize) {
+			break;
+		}
+
+		// receive incoming UDP packets
+		int len = m_UDP.read(m_Packet, sizeof(m_Packet));
+
+#ifdef DEBUG_NETWORK
+		m_Logger.trace(
+			"Received %d bytes from %s, port %d",
+			packetSize,
+			m_UDP.remoteIP().toString().c_str(),
+			m_UDP.remotePort()
+		);
+		m_Logger.traceArray("UDP packet contents: ", m_Packet, len);
+#endif
+
+		// Handshake is different, it has 3 in the first byte, not the 4th, and data
+		// starts right after
+		if (m_Packet[0] == PACKET_HANDSHAKE) {
+			if (strncmp((char*)m_Packet + 1, "Hey OVR =D 5", 12) != 0) {
+				m_Logger.error("Received invalid handshake packet");
+				continue;
+			}
+
+			m_ServerHost = m_UDP.remoteIP();
+			m_ServerPort = m_UDP.remotePort();
+			m_LastPacketTimestamp = millis();
+			m_Connected = true;
+
+			statusManager.setStatus(SlimeVR::Status::SERVER_CONNECTING, false);
+			ledManager.off();
+
+			m_Logger.debug(
+				"Handshake successful, server is %s:%d",
+				m_UDP.remoteIP().toString().c_str(),
+				m_UDP.remotePort()
+			);
+
+			break;
+		}
+	}
+
+	auto now = millis();
+
+	// This makes the LED blink for 20ms every second
+	if (m_lastConnectionAttemptTimestamp + 1000 < now) {
+		m_lastConnectionAttemptTimestamp = now;
+		m_Logger.info("Searching for the server on the local network...");
+		Connection::sendTrackerDiscovery();
+		ledManager.on();
+	} else if (m_lastConnectionAttemptTimestamp + 20 < now) {
+		ledManager.off();
+	}
+}
+
+void Connection::reset() {
+	m_UDP.begin(m_ServerPort);
+	m_Connected = false;
+
+	statusManager.setStatus(SlimeVR::Status::SERVER_CONNECTING, true);
+}
+
+void Connection::update() {
+	auto sensor1 = sensorManager.getFirst();
+	auto sensor2 = sensorManager.getSecond();
+
+	if (!m_Connected) {
+		searchForServer();
+	}
+
+	int packetSize = m_UDP.parsePacket();
+	if (!packetSize) {
+		return;
+	}
+
+	m_LastPacketTimestamp = millis();
+	int len = m_UDP.read(m_Packet, sizeof(m_Packet));
+
+#ifdef DEBUG_NETWORK
+	m_Logger.trace(
+		"Received %d bytes from %s, port %d",
+		packetSize,
+		m_UDP.remoteIP().toString().c_str(),
+		m_UDP.remotePort()
+	);
+	m_Logger.traceArray("UDP packet contents: ", m_Packet, len);
+#endif
+
+	switch (convert_chars<int>(m_Packet)) {
+		case PACKET_RECEIVE_HEARTBEAT:
+			sendHeartbeat();
+			break;
+
+		case PACKET_RECEIVE_VIBRATE:
+			break;
+
+		case PACKET_RECEIVE_HANDSHAKE:
+			// Assume handshake successful
+			m_Logger.warn("Handshake received again, ignoring");
+			break;
+
+		case PACKET_RECEIVE_COMMAND:
+			break;
+
+		case PACKET_CONFIG:
+			break;
+
+		case PACKET_PING_PONG:
+			returnLastPacket(len);
+			break;
+
+		case PACKET_SENSOR_INFO:
+			if (len < 6) {
+				m_Logger.warn("Wrong sensor info packet");
+				break;
+			}
+
+			if (m_Packet[4] == sensor1->getSensorId()) {
+				m_AckedSensorState1 = m_Packet[5];
+			} else if (m_Packet[4] == sensor2->getSensorId()) {
+				m_AckedSensorState2 = m_Packet[5];
+			}
+
+			break;
+	}
```

Comment:
`packetSize` will be frozen to 0 when disconnected, thus returning here skips the timeout detection completely. This patch will only skip the UDP packet parsing when `packetSize == 0` and could detect timeout correctly.
```suggestion
	if (!m_Connected) {
		searchForServer();
		return;
	}

	int packetSize = m_UDP.parsePacket();
	if (packetSize) {
		m_LastPacketTimestamp = millis();
		int len = m_UDP.read(m_Packet, sizeof(m_Packet));

		#ifdef DEBUG_NETWORK
		m_Logger.trace(
			"Received %d bytes from %s, port %d",
			packetSize,
			m_UDP.remoteIP().toString().c_str(),
			m_UDP.remotePort()
		);
		m_Logger.traceArray("UDP packet contents: ", m_Packet, len);
		#endif

		switch (convert_chars<int>(m_Packet)) {
			case PACKET_RECEIVE_HEARTBEAT:
				sendHeartbeat();
				break;

			case PACKET_RECEIVE_VIBRATE:
				break;

			case PACKET_RECEIVE_HANDSHAKE:
				// Assume handshake successful
				m_Logger.warn("Handshake received again, ignoring");
				break;

			case PACKET_RECEIVE_COMMAND:
				break;

			case PACKET_CONFIG:
				break;

			case PACKET_PING_PONG:
				returnLastPacket(len);
				break;

			case PACKET_SENSOR_INFO:
				if (len < 6) {
					m_Logger.warn("Wrong sensor info packet");
					break;
				}

				if (m_Packet[4] == sensor1->getSensorId()) {
					m_AckedSensorState1 = m_Packet[5];
				} else if (m_Packet[4] == sensor2->getSensorId()) {
					m_AckedSensorState2 = m_Packet[5];
				}

				break;
		}
	}
```

Reply:
I think I solved this properly